validator/src/validator.rs

```diff
@@ -413,19 +424,105 @@ impl Validator {
         }
     }
 
-    pub fn on_pbft_proposal(&self, hash: &Blake2bHash) {
-        let state = self.state.write();
+    pub fn on_pbft_proposal(&self, proposal: &PbftProposal) {
+        // NOTE: Upto this point only the header was checked
+
+        let hash: Blake2bHash = proposal.header.hash();
         trace!("Received proposal: {}", hash);
-        // View change messages should only be sent by active validators.
+
+        let mut state = self.state.write();
+
+        // Signed proposal messages should only be sent to active validators.
         if state.status != ValidatorStatus::Active {
             return;
         }
 
+        // Check if the proposal is for the current view change. Note that the ValidatorNetwork
+        // checks that the proposal is valid upto the last block we know, but can't check against
+        // the view number that the validator tracks.
+        if proposal.header.view_number < state.view_number {
+            return;
+        }
+
+        // Check if we're already committed to the proposal
+        if state.committed {
+            debug!("Proposal, but already committed: {}", hash);
+            return;
+        }
+
+        // Check if this proposal is "better"
+        if let Some((current_proposal_hash, _)) = &state.proposal {
+            if *current_proposal_hash == hash {
+                // We either produced this block, or already started pBFT. Actually if we're not the
+                // producer this should never be called multiple times for the same hash.
+                warn!("Proposal already known: {}", hash);
+                return
```

Comment:
The code is a bit inconsistent here with `return` vs `return;`.
I'll make this consistent when rebasing it.

Reply:
Better?